### ISSUE TITLE
K8s Operator: Failed to Reconcile StatefulSet - update to fields forbidden

### ISSUE THREAD

**Issue by tutman96** [tailscale/tailscale#12889]
### What is the issue?

I am receiving the following error from the Kubernetes operator which is preventing updates of the machine versions when the operator updates

```
"failed to provision: failed to reconcile statefulset: StatefulSet.apps \"ts-XXXX-7cbzc\" is invalid: spec: Forbidden: updates to statefulset spec for fields other than 'replicas', 'ordinals', 'template', 'updateStrategy', 'persistentVolumeClaimRetentionPolicy' and 'minReadySeconds' are forbidden","stacktrace":"sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler\n\tsigs.k8s.io/controller-runtime@v0.18.4/pkg/internal/controller/controller.go:324\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem\n\tsigs.k8s.io/controller-runtime@v0.18.4/pkg/internal/controller/controller.go:261\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2\n\tsigs.k8s.io/controller-runtime@v0.18.4/pkg/internal/controller/controller.go:222"
```

I have enabled debug logs and it doesn't give any additional indication of what is being accomplished.

### Steps to reproduce

The service I am using to create a machine is:
```yaml
apiVersion: v1
kind: Service
metadata:
  annotations:
    tailscale.com/expose: "true"
  name: XXX
spec:
  type: ExternalName
  externalName: XXX # A record pointing to an internal ip
  ports:
    - port: 80
      name: http
      targetPort: 80
```

The current StatefulSet:
```yaml
apiVersion: apps/v1
kind: StatefulSet
metadata:
  creationTimestamp: '2024-05-27T17:02:18Z'
  generation: 2
  labels:
    tailscale.com/managed: 'true'
    tailscale.com/parent-resource: XXX
    tailscale.com/parent-resource-ns: XXX
    tailscale.com/parent-resource-type: svc
  name: ts-XXX-7cbzc
  namespace: tailscale
  resourceVersion: '15811672'
  uid: 97e5b09b-75ae-4b07-acc3-4f2810052e13
spec:
  persistentVolumeClaimRetentionPolicy:
    whenDeleted: Retain
    whenScaled: Retain
  podManagementPolicy: OrderedReady
  replicas: 1
  revisionHistoryLimit: 10
  selector:
    matchLabels:
      app: eb9fee29-f3f2-4074-81ae-c069a8d35b9c
  serviceName: ts-XXX-7cbzc
  template:
    metadata:
      annotations:
        tailscale.com/operator-last-set-cluster-dns-name: XXX
        tailscale.com/operator-last-set-config-file-hash: e89970c266f207b1d1e3033aad87434375b755128e9f7e465974c59e943b349b
      creationTimestamp: null
      deletionGracePeriodSeconds: 10
      labels:
        app: eb9fee29-f3f2-4074-81ae-c069a8d35b9c
        tailscale.com/managed: 'true'
        tailscale.com/parent-resource: XXX
        tailscale.com/parent-resource-ns: XXX
        tailscale.com/parent-resource-type: svc
    spec:
      containers:
        - env:
            - name: TS_USERSPACE
              value: 'false'
            - name: POD_IP
              valueFrom:
                fieldRef:
                  apiVersion: v1
                  fieldPath: status.podIP
            - name: TS_KUBE_SECRET
              value: ts-XXX-7cbzc-0
            - name: EXPERIMENTAL_TS_CONFIGFILE_PATH
              value: /etc/tsconfig/tailscaled
            - name: TS_EXPERIMENTAL_VERSIONED_CONFIG_DIR
              value: /etc/tsconfig
            - name: TS_DEBUG_FIREWALL_MODE
              value: auto
            - name: TS_EXPERIMENTAL_DEST_DNS_NAME
              value: XXX
          image: 'tailscale/tailscale:v1.70.0'
          imagePullPolicy: Always
          name: tailscale
          resources: {}
          securityContext:
            capabilities:
              add:
                - NET_ADMIN
          terminationMessagePath: /dev/termination-log
          terminationMessagePolicy: File
          volumeMounts:
            - mountPath: /etc/tsconfig
              name: tailscaledconfig
              readOnly: true
      dnsPolicy: ClusterFirst
      initContainers:
        - args:
            - >-
              sysctl -w net.ipv4.ip_forward=1 && if sysctl
              net.ipv6.conf.all.forwarding; then sysctl -w
              net.ipv6.conf.all.forwarding=1; fi
          command:
            - /bin/sh
            - '-c'
          image: 'tailscale/tailscale:v1.70.0'
          imagePullPolicy: IfNotPresent
          name: sysctler
          resources: {}
          securityContext:
            privileged: true
          terminationMessagePath: /dev/termination-log
          terminationMessagePolicy: File
      restartPolicy: Always
      schedulerName: default-scheduler
      securityContext: {}
      serviceAccount: proxies
      serviceAccountName: proxies
      terminationGracePeriodSeconds: 30
      volumes:
        - name: tailscaledconfig
          secret:
            defaultMode: 420
            secretName: ts-XXX-7cbzc-0
  updateStrategy:
    rollingUpdate:
      partition: 0
    type: RollingUpdate
status:
  availableReplicas: 1
  collisionCount: 0
  currentReplicas: 1
  currentRevision: ts-XXX-7cbzc-54b9989765
  observedGeneration: 2
  readyReplicas: 1
  replicas: 1
  updateRevision: ts-XXX-7cbzc-54b9989765
  updatedReplicas: 1
```

### Are there any recent changes that introduced the issue?

Unsure when this error started occurring, but instance wouldn't update past 1.63.1 without manually editing the statefulset, and I've stayed close to up to date on the K8s operator

### OS

Linux

### Tailscale version

1.70.0

### Other software

Kubernetes Operator on K3s

### Bug report

BUG-2636b9a23a941f8c33bffcb04c819c796c808b22d0069d1c67fd0ff5fe077619-20240722172833Z-902c839944e35db7

**Comment by irbekrm** [comment]
Hi! Could you paste in some more of the operator logs? You can increase log level like this https://tailscale.com/kb/1446/kubernetes-operator-troubleshooting#operator-logs

(We don't get access to those logs via bugreport).

I've just tried a 1.66 -> 1.70 update and couldn't reproduce this, so I'm just wondering what field update it could be that breaks. What are the manual changes that you are making to get it working?

**Comment by tutman96** [comment]
Here are the relevant debug logs
```
{"level":"debug","ts":"2024-07-22T18:34:37Z","logger":"service-reconciler","msg":"starting reconcile","service-ns":"XXX","service-name":"XXX"}
{"level":"debug","ts":"2024-07-22T18:34:37Z","logger":"service-reconciler","msg":"reconciling headless service for StatefulSet","service-ns":"XXX","service-name":"XXX"}
{"level":"debug","ts":"2024-07-22T18:34:37Z","logger":"service-reconciler","msg":"secret tailscale/ts-XXX-7cbzc-0 already exists","service-ns":"XXX","service-name":"XXX"}
{"level":"debug","ts":"2024-07-22T18:34:37Z","logger":"service-reconciler","msg":"patching the existing proxy Secret with tailscaled config {\"Version\":\"alpha0\",\"Locked\":false,\"AuthKey\":\"**redacted**\",\"Hostname\":\"XXX-XXX\",\"acceptDNS\":false,\"acceptRoutes\":false,\"NoStatefulFiltering\":false}","service-ns":"XXX","service-name":"XXX"}
{"level":"debug","ts":"2024-07-22T18:34:37Z","logger":"service-reconciler","msg":"reconciling statefulset tailscale/ts-XXX-7cbzc","service-ns":"XXX","service-name":"XXX"}
{"level":"debug","ts":"2024-07-22T18:34:37Z","logger":"service-reconciler","msg":"reconcile finished","service-ns":"XXX","service-name":"XXX"}
{"level":"error","ts":"2024-07-22T18:34:37Z","msg":"Reconciler error","controller":"service-reconciler","namespace":"XXX","name":"XXX","reconcileID":"de1ef8f0-0762-4330-8b0f-14d1f4240400","error":"failed to provision: failed to reconcile statefulset: StatefulSet.apps \"ts-XXX-7cbzc\" is invalid: spec: Forbidden: updates to statefulset spec for fields other than 'replicas', 'ordinals', 'template', 'updateStrategy', 'persistentVolumeClaimRetentionPolicy' and 'minReadySeconds' are forbidden","stacktrace":"sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler\n\tsigs.k8s.io/controller-runtime@v0.18.4/pkg/internal/controller/controller.go:324\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem\n\tsigs.k8s.io/controller-runtime@v0.18.4/pkg/internal/controller/controller.go:261\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2\n\tsigs.k8s.io/controller-runtime@v0.18.4/pkg/internal/controller/controller.go:222"}
```

The only change I made manually to the statefulset was changing the image tag to v1.70.0 to see if that fixed it, but still get the above errors.

I have the node shared outside of my tailnet so deleting and recreating it is a bit disruptive, but if there was a way to maintain its state but replace the statefulset so it reconciles correctly that would be acceptable.

**Comment by irbekrm** [comment]
Thanks for the extra info- I cannot tell which fields those might be from the logs, so will have to try to reproduce. If you remember what version you originally deployed that would be useful too.
I won't get to it tonight (European hours), but will try to see if we can add a patch or at least manual update instructions once I see what fields are changing.

**Comment by tutman96** [comment]
Thanks. I am working to get audit logs turned on in my cluster so I can see what the patch body looks like and will post that here.

**Comment by tutman96** [comment]
Okay, so here's what is changing with the update the operator is performing
```diff
spec:
  selector:
    matchLabels:
-     app: eb9fee29-f3f2-4074-81ae-c069a8d35b9c
+     app: 12d86b7e-4ce6-4111-8231-758dada1257d
  template:
    metadata:
      labels:
-       app: eb9fee29-f3f2-4074-81ae-c069a8d35b9c
+       app: 12d86b7e-4ce6-4111-8231-758dada1257d
```
which matches the uid of the Service resource. Potentially when I switched over to using the ExternalName resource there was something that got out of sync... or maybe I had done a force replace on the Service. 🤔

**Comment by irbekrm** [comment]
Thanks for the extra info!

The app label is the UID of the `Service` resource as you already found out. I think a UID of a kube object cannot change during its lifetime and we do add finalizers to the `Service`s that we manage, so it shouldn't have been possible to delete the `Service` and bypass operator's cleanup without manually removing the finalizers.
However, we could try to recover from this more gracefully and/or allow for flows where the tailscale node's state is not deleted, I'm thinking we should either force the update in the operator (maybe server side apply could help with that) or delete and recreate the `StatefulSet` only if we see that the UID no longer matches that of the `Service` (cc @maisem @tomhjp )
Out of interest, do you remeber what was the flow that you were aiming for when changing the `Service` resource?

> I have the node shared outside of my tailnet so deleting and recreating it is a bit disruptive, but if there was a way to maintain its state but replace the statefulset so it reconciles correctly that would be acceptable.

You should be able to achieve that by force applying the `StatefulSet` with the new `app` label value.

If you
- `kubectl get <sts-name>  -n tailscale -oyaml > sts.yaml`
- set `spec.selector.matchLabels[app]` to the current `Service` UID
- set `spec.template.metadata.labels[app]` to the current `Service` UID
- remove `metadata.resourceVersion`, `metadata.creationTimestamp`, `metadata.generation`, `metadata.uid`
- `kubectl apply --force=true -f sts.yaml`

That should do it- let us know if it worked.
In any case, even if it fails, it will not cause the operator to delete the tailscale node's state- that would only happen if you delete the `Service` or its tailscale annotation.

**Comment by tutman96** [comment]
Great thanks for the confirmation. I applied the changes as you suggested and now get a TailscaleProxyReady=True status in the service now 🎉  (love that new status by the way, was the reason I caught this problem in the first place)

This was quite a while ago, but I think what happened was I converted this service from a LoadBalancer to an ExternalName, which requires a recreate. I use ArgoCD for management of the resources, so I just ran a Sync with "Replace" and "Force" options enabled. I will check to see if I can reproduce this problem and report back.

**Comment by tutman96** [comment]
Just ran a replace on an existing Service (LoadBalancer) and the Tailscale node was deleted and recreated as expected. I'll just blame this on me doing something wrong in Kubernetes and maybe that plus an older version of the operator got it stuck in a weird state.

**Comment by irbekrm** [comment]
Someone did open an issue a while ago, reporting a similar problem https://github.com/tailscale/tailscale/issues/11482 when using ArgoCD. I wasn't able to reproduce it then also, but maybe Argo does sometimes remove finalizers when force replacing objects. To be investigated at some point

**Comment by tutman96** [comment]
Looking through ArgoCD source for a bit... looks like replace + force might be the culprit. From the kubectl replace docs (https://kubernetes.io/docs/reference/kubectl/generated/kubectl_replace/#options)

> --force | If true, immediately remove resources from API and bypass graceful deletion. Note that immediate deletion of some resources may result in inconsistency or data loss and requires confirmation.

So there is a probability that the operator didn't get hooked into the process during the delete and potentially was recreated by the time the operator got notified.

I was able to reproduce this problem, but it did require manually deleting the finalizer after deleting the svc resource while the operator wasn't running. I couldn't get the svc to cleanly delete without the operator running from anything I did in Argo directly. So regardless, I think this is an obscure edge case that I'm not even sure is handled by other operators.
